tools/yulPhaser/Mutations.cpp

```diff
@@ -180,3 +180,138 @@ function<Crossover> phaser::fixedPointCrossover(double _crossoverPoint)
 		return get<0>(fixedPointSwap(_chromosome1, _chromosome2, concretePoint));
 	};
 }
+
+namespace
+{
+
+ChromosomePair fixedTwoPointSwap(
+	Chromosome const& _chromosome1,
+	Chromosome const& _chromosome2,
+	size_t _crossoverPoint1,
+	size_t _crossoverPoint2
+)
+{
+	assert(_crossoverPoint1 <= _chromosome1.length());
+	assert(_crossoverPoint1 <= _chromosome2.length());
+	assert(_crossoverPoint2 <= _chromosome1.length());
+	assert(_crossoverPoint2 <= _chromosome2.length());
+
+	size_t lowPoint = min(_crossoverPoint1, _crossoverPoint2);
+	size_t highPoint = max(_crossoverPoint1, _crossoverPoint2);
+
+	auto begin1 = _chromosome1.optimisationSteps().begin();
+	auto begin2 = _chromosome2.optimisationSteps().begin();
+	auto end1 = _chromosome1.optimisationSteps().end();
+	auto end2 = _chromosome2.optimisationSteps().end();
+
+	return {
+		Chromosome(
+			vector<string>(begin1, begin1 + lowPoint) +
+			vector<string>(begin2 + lowPoint, begin2 + highPoint) +
```

Comment:
Is `_chromosomeX.length() == _chromosomeX.optimisationSteps().length()`?
I'm just wondering if `begin2 + highPoint` could go past `length`, since I don't see anything explicit avoiding that. Maybe add an assertion?

Reply:
> Is `_chromosomeX.length() == _chromosomeX.optimisationSteps().length()`?

Yes. `length()` is just a convenience helper that returns `_chromosomeX.optimisationSteps().size()`.
I was actually thinking about adding a similar helper to `Population` because `Population.individuals().size()` gets used a lot, especially in tests.

> I'm just wondering if `begin2 + highPoint` could go past `length`, since I don't see anything explicit avoiding that. Maybe add an assertion?

There are a few assertions above. At the beginning of the function I assert that both points are within the length of both chromosomes. Which basically means max(point1, point2) <= min(length1, length2). And `highPoint` is simply max(point1, point2).